Serialiser_Engine: IsValidJson checks if string is in valid format

desaiwangBH opened the pull request:
<!-- PLEASE ENSURE YOU REVIEW THE CONTENT OF EACH PR CAREFULLY, INCLUDING SUBSEQUENT COMMENTS BY YOURSELF OR OTHERS. -->
<!-- IN PARTICULAR PLEASE ENSURE THAT SENSITIVE OR INAPPROPRIATE INFORMATION IS NOT UPLOADED -->

### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #1427  

<!-- Add short description of what has been fixed -->

- checks if json is valid
- only checks if formatting is correct, does not validate schema

### Test files
<!-- Link to test files to validate the proposed changes -->


### Changelog
<!-- Text to go into changelog if applicable -->
<!-- Please see https://github.com/BHoM/documentation/wiki/changelog for guidelines -->


### Additional comments
<!-- As required -->
As per Fraser's suggestions, creating a new method inside Query that is similar to Convert/ToJson to check if json is valid. The method returns true if the input string is a valid json, false otherwise.

FraserGreenroyd commented on Serialiser_Engine/Query/Json.cs:
```diff
@@ -0,0 +1,63 @@
+        {
+            return Regex.IsMatch(str, @"\,(?!\s*?[\{\[\""\w])");
+        }
+
```
```suggestion
```

FraserGreenroyd commented on Serialiser_Engine/Query/Json.cs:
```diff
@@ -0,0 +1,63 @@
+        //JSON validation based on code from https://github.com/prototypejs/prototype/blob/560bb59414fc9343ce85429b91b1e1b82fdc6812/src/prototype/lang/string.js#L699
+        
+        [Description("Checks if a string is a valid Json")]
+        [Input("string", "String to check.")]
```
```suggestion
        [Input("str", "String to check validity of.")]
```

FraserGreenroyd commented on Serialiser_Engine/Query/Json.cs:
```diff
@@ -0,0 +1,63 @@
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        //JSON validation based on code from https://github.com/prototypejs/prototype/blob/560bb59414fc9343ce85429b91b1e1b82fdc6812/src/prototype/lang/string.js#L699
```
```suggestion
```

FraserGreenroyd commented on Serialiser_Engine/Query/Json.cs:
```diff
@@ -0,0 +1,63 @@
+        [Input("string", "String to check.")]
+        [Output("validity", "True if string is a json, false otherwise.")]
+        public static bool IsValidJson(this string str)
+        {
```
```suggestion
        {
            //JSON validation based on code from https://github.com/prototypejs/prototype/blob/560bb59414fc9343ce85429b91b1e1b82fdc6812/src/prototype/lang/string.js#L699
```

desaiwangBH commented on Serialiser_Engine/Query/Json.cs:
```diff
@@ -0,0 +1,63 @@
+
+        //JSON checking trailing coomas regex based on: https://stackoverflow.com/questions/34344328/json-remove-trailing-comma-from-last-object
+        private static bool HasTrailingCommas(this string str)
+        {
```
commas spelled wrong, I updated as suggested on my local branch and pushed to origin

adecler reviewed (commented):
A few things from the top of my head that are not valid Json and yet are returning true:
 - `"A"`
 - `[{`
 - `[{:}{,}]`

desaiwangBH commented:
> A few things from the top of my head that are not valid Json and yet are returning true:
> 
> * `"A"`
> * `[{`
> * `[{:}{,}]`

@adecler 
According to [JsonLint](https://jsonlint.com/), `"A"` is a valid json. However, `A` would is not a valid json, but that is already returning false.

As for the rest of the cases, I realized that there was some issues with the regex. So instead I reverted the IsValidJson method logic back to using BSON, while adding a check that does return single values as valid ("A", true, null, 3, for example). However, I also noticed that BSON has some intelligence when parsing, so for [{\"name\": \"John\", city:"NYC"}] IsValidJson will return true even though city is not inside quotes/is not a string. 

I argue this is ok, since the `FromJson` also uses BSON methods, so all jsons that are determined to be valid can be parsed.

Also added unit tests. Let me know if there's anything else to do.

FraserGreenroyd commented:
Following offline discussion with @adecler and @desaiwangBH it has been agreed that this PR isn't needed anymore with the following reasons provided:

 - The Serialiser_Engine has come along a fair bit since January 2020 when #1427 was originally raised
 - The Adapters have likewise undergone improvement following work done by @alelom since January 2020
 - The JSON format used by BHoM, while needing to be valid JSON, also needs to be valid objects to deserialise to, as such `"A"`, while valid JSON, is not going to be able to be parsed by BHoM, so saying it is valid or invalid JSON doesn't help matters
 
Thanks @desaiwangBH for the research into the issue, we'll keep the code on these commits should we change our mind in the future, but as of today, this is the decision reached as a collective 😄
